src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java

```diff
@@ -0,0 +1,160 @@
+                return null;
+            }
+            double distance = getDistance(travelerPosition.currentPosition, new Coordinates(firstStep));
+            return (distance <= TRIP_INSTRUCTION_UPCOMING_DISTANCE)
+                ? new TripInstruction(distance, firstStep)
+                : null;
+        }
+
+        Coordinates destination = new Coordinates(travelerPosition.expectedLeg.to);
+        double distanceToDestination = getDistance(travelerPosition.currentPosition, destination);
+        if (distanceToDestination <= TRIP_INSTRUCTION_UPCOMING_DISTANCE) {
+            // Assuming traveler is approaching the destination.
+            return new TripInstruction(distanceToDestination, travelerPosition.expectedLeg.to.name);
+        }
+
+        int nearestStepIndex = getNearestStep(travelerPosition.expectedLeg, travelerPosition.currentPosition);
+        Step nearestStep = getStep(travelerPosition.expectedLeg.steps, nearestStepIndex);
+        if (nearestStep == null) {
+            return null;
+        }
+
+        Coordinates nearestStepCoords = new Coordinates(nearestStep);
+
+        if (isLastStep(nearestStepIndex, travelerPosition.expectedLeg)) {
+            // Last step.
+            if (isStepNearerToDestination(nearestStepCoords, travelerPosition.currentPosition, destination)) {
+                // Assuming traveler is approaching the last step.
+                return new TripInstruction(
+                    getDistance(travelerPosition.currentPosition, nearestStepCoords),
+                    getStep(travelerPosition.expectedLeg.steps, nearestStepIndex)
+                );
+            }
+        } else {
+            Step nextStep = travelerPosition.expectedLeg.steps.get(nearestStepIndex + 1);
+            Coordinates nextStepCoordinates = new Coordinates(nextStep);
+            if (!isStepNearerToDestination(nearestStepCoords, travelerPosition.currentPosition, nextStepCoordinates)) {
+                // Assuming traveler is passed the nearest step.
+                return new TripInstruction(
+                    getDistance(travelerPosition.currentPosition, nextStepCoordinates),
+                    nextStep
+                );
+            } else {
+                // Assuming traveler is approaching (or on) the nearest step.
+                return new TripInstruction(
+                    getDistance(travelerPosition.currentPosition, nearestStepCoords),
+                    getStep(travelerPosition.expectedLeg.steps, nearestStepIndex)
+                );
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Does the step index reference the last step in the leg.
+     */
+    private static boolean isLastStep(int stepIndex, Leg expectedLeg) {
+        return stepIndex == (expectedLeg.steps.size() - 1);
+    }
+
+    /**
+     * Determine if the step is near to the destination than the traveler.
+     */
+    private static boolean isStepNearerToDestination(
```

Comment:
This might be tricky. When you have to go around an impassable place to reach your destination, this method will think you are walking in the wrong direction. Would it be easier to just reference the step order within the leg?

Reply:
@binh-dam-ibigroup The method name is a bit misleading so I have updated it to `isFirstPositionNearer`. I have also moved the method into the `GeometryUtils` class. The trip destination is only used if working if the last step, in all other cases the next step in the leg is used i.e. the step order within the leg.